ls1intum/Artemis
# [Feature] Persist plagiarism results

### Checklist

### Motivation and Context
Currently, there are three REST endpoints to start the automated plagiarism detection for modeling, programming, and text exercises:

* `GET /modeling-exercises/{exerciseId}/check-plagiarism`
* `GET /programming-exercises/{exerciseId}/check-plagiarism`
* `GET /text-exercises/{exerciseId}/check-plagiarism`

The `/check-plagiarism` endpoints for programming and text exercises return a `TextPlagiarismResult` and the `/check-plagiarism` endpoint for modeling exercises returns a `ModelingPlagiarismResult`.

However, the plagiarism detection results are not persisted on the Artemis server and get lost if the instructor refreshes the browser page. 

The goal of this PR is to persist plagiarism detection results to avoid unnecessary reruns. Instructors should have access to the latest plagiarism results only. They can rerun the automated plagiarism detection at any time to generate new results.

### Description
First, this PR introduces a new `PlagiarismService` that offers CRUD functionality for plagiarism results using the `PlagiarismResultRepository` and `PlagiarismComparisonRepository`.

Second, the `/check-plagiarism` route handlers now use the new `PlagiarismService` to store the latest result in the database before returning them to the client.
 
Third, we added three new REST endpoints to the server to access the latest plagiarism results for modeling, programming, and text exercises respectively:

* `GET /modeling-exercises/{exerciseId}/plagiarism-result`
* `GET /programming-exercises/{exerciseId}/plagiarism-result`
* `GET /text-exercises/{exerciseId}/plagiarism-result`

Last, we added a new REST endpoint to update the status of a single PlagiarismComparison. We also adapted the client to call this endpoint whenever the user clicks _Confirm_ or _Deny_ from the plagiarism inspector (see Screenshot).

### Steps for Testing
Make sure to test the following for modeling, programming, and text exercises. You'll new a few similar submissions for each exercise type to properly test this PR.

1. Log in to Artemis
2. Navigate to Course Administration
3. Go to the Exercises of a course
4. Go to the details page of a (modeling | programming | text) exercise with similar submissions
5. Click 'Check Plagiarism'

You're now on the Plagiarism Detection page. Make sure the following works:

- The plagiarism detection page should load the latest plagiarism detection result on page load (if any exists; otherwise empty page)
- Click 'Detect plagiarism' to start a new plagiarism detection run and receive results
- Highlighting of similarities for text and programming exercises should work
- Clicking _Confirm_ or _Deny_ correctly updates the status of a single comparison
- The updated status of a comparison is persisted and loaded correctly after refreshing the page
- Clicking 'Rerun plagiarism detection' creates a new result that overwrites the previous one. Especially check that the status of each comparison is reset

### Test Coverage
<img width="997" alt="Screenshot 2021-03-23 at 15 31 15" src="https://user-images.githubusercontent.com/21973908/112162988-deb6d980-8bec-11eb-9384-c491aeaea418.png">


### Screenshots
<img width="1165" alt="Screenshot 2021-03-16 at 09 18 52" src="https://user-images.githubusercontent.com/21973908/111277337-a5083080-8638-11eb-8ed2-8777919ae182.png">
